### PR TITLE
Fps sample

### PR DESCRIPTION
**Pull request by EternalWind**
Finally, the FPS sample has been done.

### CODE REVIEW

File: engine/src/Graphics/DisplayManager.cpp [hunk 1]
```diff
@@ -64,6 +66,13 @@ void DisplayManager::Render() {
     if(mOgreRoot != nullptr && mOgreRoot->isInitialised()) {
         mOgreRoot->renderOneFrame();
         Ogre::WindowEventUtilities::messagePump();
+
+        if(mMainCamera) {
```
Comment (svenstaro) [inline engine/src/Graphics/DisplayManager.cpp]: Come on, don't do this. Compare it against nullptr or something. Also this whole thing needs to be moved somewhere where it belongs. This is not a rendering operation.

File: engine/src/Graphics/DisplayManager.cpp [hunk 2]
```diff
@@ -11,8 +11,10 @@
 #include <Core/Root.hpp>
 #include <Utils/LogManager.hpp>
 #include <Input/InputManager.hpp>
+#include <Scene/Node.hpp>
 
 #include <SFML/Window/VideoMode.hpp>
+#include <SFML/Audio/Listener.hpp>
```
Comment (svenstaro) [inline engine/src/Graphics/DisplayManager.cpp]: Try to put this somewhere else. We actually don't want Audio stuff mixed with Display stuff.

File: engine/src/Gui/GuiManager.cpp [hunk 3]
```diff
@@ -19,7 +19,7 @@
 GuiManager::GuiManager()
     : mGuiSystem(nullptr),
       mPlatform(nullptr),
-      mMouseCursorVisible(false),
+      mMouseCursorVisible(true),
```
Comment (svenstaro) [inline engine/src/Gui/GuiManager.cpp:22]: Why this change?

File: engine/src/Logic/AdvancedPlayerComponent.cpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    btScalar character_width = 0.44;
+    btConvexShape* capsule = new btCapsuleShape(character_width, character_height);
+
+    mBtGhostObject = new btPairCachingGhostObject();
+    mBtGhostObject->setWorldTransform(start_trans);
+    mBtGhostObject->setCollisionShape(capsule);
+    mBtGhostObject->setCollisionFlags(btCollisionObject::CF_CHARACTER_OBJECT);
+    /////////////////////////////////////////////////////////////////////////////////////////////////
+    //TODO: Make a better way to distinguish between AdvancedPlayerComponent and PhysicsBodyComponent.
+    //For now, it's just a null pointer check to do this.
```
Comment (svenstaro) [inline engine/src/Logic/AdvancedPlayerComponent.cpp:53]: Meh, can't you think of a better way? :/

File: engine/src/Logic/AdvancedPlayerComponent.cpp [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    if(!QObject::connect(InputManager::Get(), SIGNAL(sMouseReleased(const OIS::MouseEvent&, OIS::MouseButtonID)), 
+        this, SLOT(_HandleMouseUp(const OIS::MouseEvent&, OIS::MouseButtonID)))) {
+            Logger::Get().Error("Cannot connect the mouse moved signal with " + GetName()
+                + "'s mouse input handling slot.");
+    }
+}
+
+void AdvancedPlayerComponent::OnDestroy() {
+    if(mBtController)
+        delete mBtController;
```
Comment (svenstaro) [inline engine/src/Logic/AdvancedPlayerComponent.cpp]: Why not use shared ptr?

File: engine/src/Logic/AdvancedPlayerComponent.cpp [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    }
+    if(!QObject::connect(InputManager::Get(), SIGNAL(sMouseReleased(const OIS::MouseEvent&, OIS::MouseButtonID)), 
+        this, SLOT(_HandleMouseUp(const OIS::MouseEvent&, OIS::MouseButtonID)))) {
+            Logger::Get().Error("Cannot connect the mouse moved signal with " + GetName()
+                + "'s mouse input handling slot.");
+    }
+}
+
+void AdvancedPlayerComponent::OnDestroy() {
+    if(mBtController)
```
Comment (svenstaro) [inline engine/src/Logic/AdvancedPlayerComponent.cpp]: Please don't check like this. It's ambiguous.

File: engine/src/Logic/AdvancedPlayerComponent.cpp [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        delete mBtController;
+    if(mBtGhostObject)
+        delete mBtGhostObject;
+}
+
+void AdvancedPlayerComponent::OnEnable() {
+    SetWASDEnabled(true);
```
Comment (svenstaro) [inline engine/src/Logic/AdvancedPlayerComponent.cpp]: You need a more generic name for this. What if you rebind the keys to something different? At the very least, rename this function.

File: engine/src/Graphics/DisplayManager.cpp (hunk 2 above)
Comment (EternalWind) [inline engine/src/Graphics/DisplayManager.cpp]: So where?

File: engine/src/Graphics/DisplayManager.cpp (hunk 1 above)
Comment (EternalWind) [inline engine/src/Graphics/DisplayManager.cpp]: Why not? if can be used like this. And where to move?

File: engine/src/Gui/GuiManager.cpp (hunk 3 above)
Comment (EternalWind) [inline engine/src/Gui/GuiManager.cpp:22]: Or you will find you can never hide the cursor.

File: engine/src/Logic/AdvancedPlayerComponent.cpp (hunk 4 above)
Comment (EternalWind) [inline engine/src/Logic/AdvancedPlayerComponent.cpp:53]: I remember you told me to use nullptr-check for now.

**Comment by svenstaro** [comment]
Did you check for memory leaks? I think you might be leaking to due unprotected calls to new that don't go into a safe container on the other side.

File: engine/src/Logic/AdvancedPlayerComponent.cpp (hunk 4 above)
Comment (svenstaro) [inline engine/src/Logic/AdvancedPlayerComponent.cpp:53]: Yes, temporarily. However, now that you want to merge perhaps we can work something better out?

Comment (EternalWind) [inline engine/src/Logic/AdvancedPlayerComponent.cpp:53]: So your ideas?

**Comment by EternalWind** [comment]
Why do you think so? The only posibility of memory leaks is the bullet can't be destroy and this will only happen when it hits nothing. Once we have skybox or so to wrap the whole scene then it will never happen.

File: samples/fps/src/FPS.cpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+#include <Logic/SimplePlayerComponent.hpp>
+#include <Physics/PhysicsBodyComponent.hpp>
+#include <Logic/AdvancedPlayerComponent.hpp>
+#include <Physics/PhysicsManager.hpp>
+#include <Core/ResourceManager.hpp>
+
+#include <OgreProcedural.h>
+
+void Main::OnInitialize() {
+    dt::ResourceManager::Get()->AddDataPath(QDir("fps/data"));
```
Comment (shua) [inline samples/fps/src/FPS.cpp]: right here, just change "fps/data" to "../samples/fps/data" I assume that will work on any system.

Comment (EternalWind) [inline samples/fps/src/FPS.cpp]: But won't Qt assumes that it's current directory + "fps/data" if what I pass to it is not an absolute path?

Comment (EternalWind) [inline samples/fps/src/FPS.cpp]: Btw, if you want to run it using the binary you should put the fps sample's data folder under the same folder as the binary files. 

P.S. Remember to put the files under the engine's data folder to your data folder as well. Because some textures I used are from the engine.

Comment (shua) [inline samples/fps/src/FPS.cpp]: Qt works off of where you call the binary, not where the binary is stored. To show this, don't put the fps/data folder in bin. Instead, just cd into the samples folder and run "../bin/sample_fps". It will run without any errors, because fps/data is called relative to the samples folder. If you change the dir above to ../samples/fps/data you will be able to run the binary in the bin directory.

 Also, the engine's data folder is already added as a data path, otherwise line 19 and 20 below wouldn't work.

File: samples/CMakeLists.txt [hunk 9]
```diff
@@ -4,12 +4,14 @@ file(GLOB_RECURSE samples_server_HEADERS chat/src/server/*.hpp)
 file(GLOB_RECURSE simplepong_HEADERS simplepong/src/*.hpp)
 file(GLOB_RECURSE advancedpong_HEADERS advancedpong/src/*.hpp)
 file(GLOB_RECURSE jumptape_HEADERS jumptape/src/*.hpp)
+file(GLOB_RECURSE sample_fps_HEADERS fps/src/*.hpp)
```
Comment (shua) [inline samples/CMakeLists.txt]: Not really a biggy, but all the other HEADER names don't start with "sample". Just my ocd kicking in. All you'd have to change are this and line 14.

File: engine/src/Graphics/DisplayManager.cpp [hunk 10]
```diff
@@ -11,6 +11,7 @@
 #include <Core/Root.hpp>
 #include <Utils/LogManager.hpp>
 #include <Input/InputManager.hpp>
+#include <Scene/Node.hpp>
```
Comment (svenstaro) [inline engine/src/Graphics/DisplayManager.cpp]: You don't need this anymore.

File: engine/src/Scene/Node.cpp [hunk 11]
```diff
@@ -186,7 +197,15 @@ void Node::LookAt(Ogre::Vector3 target, Ogre::Vector3 front_vector, RelativeTo r
 void Node::SetParent(Node* parent) {
     if(parent != nullptr) {
         if(parent->FindChildNode(mName, false) == nullptr) { // we are not already a child of the new parent
-            parent->AddChildNode(this);
+            if(mParent != nullptr) {                         // Remove it from its original parent.
+                auto iter = mParent->mChildren.find(mName);
+                parent->mChildren.insert(mName, mParent->mChildren.release(iter).release());
+                mParent = parent;
+                //parent->mChildren.transfer(parent->mChildren.begin(), iter, mParent->mChildren);
```
Comment (svenstaro) [inline engine/src/Scene/Node.cpp]: Why is this commented?

Comment (EternalWind) [inline engine/src/Scene/Node.cpp]: Because transfer doesn't match the need here.

Comment (svenstaro) [inline engine/src/Scene/Node.cpp]: Then it should be removed/replaced.

File: samples/fps/src/main.cpp [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+// text, please see the LICENSE file in the root of this project or at
+// http://www.gnu.org/licenses/lgpl.html
+// ----------------------------------------------------------------------------
+
+#include "FPS.hpp"
+
+#include <Scene/Game.hpp>
+
+int main(int argc, char** argv) {
+    dt::Game game;
```
Comment (shua) [inline samples/fps/src/main.cpp:14]: dt::Root::GetInstance();
dt::InputManager::Get()->SetJailInput(true);

This will work, but you'll have to make sure to bind a key in-game for closing the game, otherwise it'll be really hard to get out. I'm working on making it work better in the engine, though. I'll wait until sven pulls, because it'll be easier.
